users/views.py

```diff
@@ -110,40 +115,44 @@ def login_process(request):
 
 def signup_view(request):
 	form = SignupForm()
-	context = {'form': form, 'result': request.GET.get('result')}
+	context = generate_signup_form(form)
+	context['result'] = request.GET.get('result')
 	return render(request, 'users/signup.html', context)
 
+def generate_signup_form(form):
+	(question, answer, chelp) = makeCaptcha()
+	return {'form': form, 'captcha': question, 'token': hash2(answer), 'captcha_help': chelp}
+
+def hash2(inp):
+	return sha256(str(inp).encode()).hexdigest()
+
 def signup_process(request):
 	form = SignupForm(request.POST)
 	if(form.is_valid()):
 
 		data_valid = True
+		captcha_input = re.sub(r'\s','',request.POST['captcha'])
+		captcha_target = re.sub(r'\s','',request.POST['captcha-token'])
 
-		# Validate captcha
-		import requests
-		import os
-		data = {
-			'secret': os.environ['GCAPTCHA_SECRET'],
-			'response': request.POST['g-recaptcha-response'],
-		}
-		r = requests.post('https://www.google.com/recaptcha/api/siteverify', data=data)
-
-		if not r.json()['success']:
-			form.add_error('g-recaptcha-response', ValidationError('Captcha failed.'))
-			data_valid = False
+		#print(hash2(captcha_input), captcha_input, captcha_target)
 
+		# CHECK CAPTCHA!
+		if hash2(captcha_input) != captcha_target:
+			# Should be attached to captcha, but this is the closest thing.
+			form.add_error('username', ValidationError('You didn\'t answer the captcha correctly!'))
```

Comment:
might be worth changing it so that the captcha is part of the form
otherwise, it'd make more sense to be a non-field error (`add_error(None, ...`)

Reply:
Done :ear_of_rice: